### PR TITLE
[Google Blockly] Address RTL workspace scrolling issues

### PR DESCRIPTION
**Pull request by mikeharv**
During the most recent bug bash for Music Lab, two bugs were identified which related to workspace scrolling in RTL. First, when dragging a block from right-side toolbox, the workspace scrolls away:

https://github.com/user-attachments/assets/c9eecbba-5614-4a24-9608-fd0b736495f7

Second, resizing the window scrolls the workspace unexpectedly:

https://github.com/user-attachments/assets/de896e0f-9ce5-43ef-8e78-e7a75d9828bc

The first bug had a relatively clear fix: https://github.com/code-dot-org/code-dot-org/pull/62206/commits/642cfecf28054469a02e9da828994c317166d91b

The method `getPaddedContent_` was erroneously using the same logic to find the right and left edges of the content in both LTR and RTL. Previously, this used 0 (fixed) to indicate the left, and the edge of right edge of the view port or content (ie. blocks), whichever was greater. In RTL, we need to do the opposite. That is, the right should be fixed and the left should be flexible based on the left edge of the view port or content.

The bug occurred when a block was dragged from the right-side toolbox. At that moment, the new block is considered part of the content and its right-edge was greater than the right-edge of the view port. This caused everything to shift temporarily until the block is dropped. Now, blocks can be dragged out without any glitch:

https://github.com/user-attachments/assets/4fef8ed5-b8fd-44d9-8f6a-f2f5c30e464a

The second issue was much harder to fix, and unfortunately, is actually made _worse_ once the first issue is fixed. If we start correctly calculating the right fixed edge, when the workspace scrolls on resize, Blockly will, in attempt to be helpful, bump the blocks back into bounds. (This is the normal behavior for when a user would place a block slightly out of bounds/too high.)
bump

https://github.com/user-attachments/assets/fe2d0b8c-f559-430f-ae89-7c0226bc6149

This issue really seems to stem from a core Blockly issue where the workspace is always anchored left, even in RTL. It's felt most acutely here because of us some customizations in how we handle fixed edges and scrollbars. However, this underlying issue makes finding the "right" fix impossible. An issue has been filed with Blockly:

- https://github.com/google/blockly/issues/8637

In order to alleviate these issues for RTL users, I had to employ a hacky workaround. I added an event handler for browser resizes that will move RTL blocks by the same number of workspace units as the delta of change in the viewport width. E.g., if the workspace shrinks by 50, the blocks need to move left by 50. However, this happens after Blockly would potentially bump the blocks back in bounds, so I needed to disable fixed edges as well. (Otherwise blocks would potentially move left _twice_ and end up out of whack.) In order to calculate the delta, we record the current width of the view port when the workspace finishes loading and after each change. Because this change affects all Google Blockly labs, we also need to record the previous width whenever the modal function editor is opened.

With both changes in place, blocks now appear to remain stationary when the browser is resized. A slight flutter is visible in the block's position, because we are reacting to the initial incorrect placement caused by the resize.

https://github.com/user-attachments/assets/ddd4290d-fd68-409a-925e-dd5fcbac54db

The workspace serialization is also unaffected here, which means users can save and reload their project in any size browser window and we will place the blocks correctly.

The main drawback to this approach (not counting its nature as a hack), is that fixed edges are turned off for RTL workspace. This means that users can place blocks such that they straddle the top or right edges. Because we don't allow scrolling past these points (due to how we calculate the padded content), the user would need to manually fix the position of these blocks, if desired.

Hopefully this trade-off is worth it!

## Links

- https://codedotorg.atlassian.net/browse/LABS-1180
- https://codedotorg.atlassian.net/browse/LABS-1181
- https://github.com/google/blockly/issues/8637

## Testing story

Manually tested using Music Lab, Flappy (with multiple top blocks), and Artist (modal function editor). I confirmed that projects can still be saved and reloaded and that we appropriately handle "aggressive" window resizing.

## Deployment strategy/Follow-up work

I will communicate with the Blockly team in order to determine if a fix to the underlying issue is possible. If it is, we could eventually restore our fixed edge setting and remove the event handlers.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by dju90** [comment]
Thanks so much for finding a solution here and communicating the downsides. I think this tradeoff is worth it.

**Review by molly-moen — APPROVED** [review]
Looks good to me!

**Comment by breville** [comment]
> In order to alleviate these issues for RTL users, I had to employ a hacky workaround. I added an event handler for browser resizes that will move RTL blocks by the same number of workspace units as the delta of change in the viewport width.

Did you have any luck pursuing the alternate idea we discussed, which was to adjust the scroll position after a resize, to avoid having to relocate the blocks themselves?

**Comment by ebeastlake** [comment]
@breville I paired with Mike on this yesterday while he was trying to get that approach to work, so I think the answer is yes, but I'll let Mike respond next week :)

**Comment by mikeharv** [comment]
> > In order to alleviate these issues for RTL users, I had to employ a hacky workaround. I added an event handler for browser resizes that will move RTL blocks by the same number of workspace units as the delta of change in the viewport width.
> 
> Did you have any luck pursuing the alternate idea we discussed, which was to adjust the scroll position after a resize, to avoid having to relocate the blocks themselves?

I did pursue it, for most of a day, but ultimately it wasn't fruitful. While I did find a way to manually scroll the workspace, the problem was that the metrics for the scrollbar wouldn't support what we imagined doing. For example, if the workspace was made larger, moving the blocks towards the center, there was no scrollbar and no detectable workspace to the left of the viewable area to scroll into.

When we chatted, I had some vague worries that moving the blocks could negatively impact serialization, but it appears that was unfounded. I have not found any other viable solutions, and I think that the tradeoffs presented here are worth moving forward.

File: apps/src/blockly/addons/cdoMetricsManager.ts [hunk 1]
```diff
@@ -26,7 +25,16 @@ export default class MetricsManager extends ScrollMetricsManager {
       viewMetrics.height
     );
 
-    const right = Math.max(contentRight, viewMetrics.width);
+    const blockMargin = 20;
```
Comment (breville) [inline apps/src/blockly/addons/cdoMetricsManager.ts:29]: Quick comment describing this?

**Review by breville — APPROVED** [review]
Thanks for digging deep into these RTL issues!

> The workspace serialization is also unaffected here, which means users can save and reload their project in any size browser window and we will place the blocks correctly.

This is a big relief, and as a result it sounds like resizing won't force an additional save.  It would be great if blockly used offset from the right in its internal representation, as it does for its serialization.  Given the current state of affairs, your changes here seem like a great solution for now.

**Pull request merged**